### PR TITLE
[Bug 1052510] Should not possible to create page as child of redirect

### PR DESCRIPTION
**Pull request by safwanrahman**
This is code is actually based on kitsune.
I could not find any better way to handle this other than this way.
Copied and refactored and ported for kuma
@jwhitlock @escattone r?

### CODE REVIEW

**Comment by jwhitlock** [comment]
This works well in my local development environment. I created some pages, and then used the "move page" feature to make them into redirects:

* /en-US/docs/user:jwhitlock/redirect1 to /en-US/docs/redirect1
* /en-US/docs/user:jwhitlock/redirect2 to /en-US/docs/Mozilla/Add-ons/redirect2 (zone URL)

I then attempted to create two new pages, and they were converted on the move page view.

* /en-US/docs/user:jwhitlock/redirect1/subpage converted to /en-US/docs/redirect1/subpage
* /en-US/docs/user:jwhitlock/redirect2/subpage converted to /en-US/docs/Mozilla/Add-ons/redirect2/subpage

So, we're 1/3 of the way there 😄.  We need a view test of a zoned URL, and unit tests w/ coverage for the new methods (``Document.from_url()``, ``Document.get_redirect_document()``, ``get_doc_components_from_url``.  There may be existing tests in kitsune.  Bonus points for converting them to py.test native tests (``assert x==y`` instead of ``self.assertEqual(x, y)`` or ``_eq(x, y)``.

Detailed code review to follow.

File: kuma/wiki/models.py [hunk 1]
```diff
@@ -1359,6 +1400,14 @@ def get_redirect_url(self):
                 elif len(url) == 1 and url[0] == '/':
                     return url
 
+    def get_redirect_document(self, id_only=False):
```
Comment (jwhitlock) [inline kuma/wiki/models.py]: ``id_only`` is ``True`` for the only caller, and should be the default.

File: kuma/wiki/models.py [hunk 2]
```diff
@@ -380,6 +380,47 @@ def get_summary_html(self, *args, **kwargs):
     def get_summary_text(self, *args, **kwargs):
         return self.get_summary(strip_markup=True)
 
+    @classmethod
+    def from_url(cls, url, required_locale=None, id_only=False,
```
Comment (jwhitlock) [inline kuma/wiki/models.py]: ``id_only`` is ``True`` for the only caller, and should be the default.

File: kuma/wiki/models.py [hunk 3]
```diff
@@ -380,6 +380,47 @@ def get_summary_html(self, *args, **kwargs):
...
+                 check_host=True):
```
Comment (jwhitlock) [inline kuma/wiki/models.py]: I can see how ``required_locale`` and ``check_host`` might be useful in the future, but not in the context of this bug.  I'd lean toward removing them and the alternate code paths, and maybe linking to the same function in the kitsune repo, so they can be added back if the need arises.

File: kuma/wiki/tests/test_views.py [hunk 4]
```diff
@@ -1190,6 +1190,22 @@ def test_create_on_404(self):
...
         eq_(302, response.status_code)
 
+    def test_creating_child_of_redirect(self):
+        """While try to create a child of a redirect,
+        the parent of the child should be redirect's parent"""
+        self.client.login(username='admin', password='testpass')
+        rev = revision(is_approved=True, save=True)
+        doc = rev.document
+        child_full_slug = doc.slug + "/" + "children_document"
+        doc._move_tree(new_slug="moved_doc")
```
Comment (jwhitlock) [inline kuma/wiki/tests/test_views.py]: ``_move_tree`` is such a complex way to create the redirect.  Manually creating the two pages would be faster.  A redirect is just a page where the content looks like ``REDIRECT_CONTENT``.

File: kuma/wiki/tests/test_views.py [hunk 5]
```diff
@@ -1190,6 +1190,22 @@ def test_create_on_404(self):
...
+        doc.refresh_from_db()
+        url = reverse('wiki.document', args=[child_full_slug])
+        response = self.client.get(url, follow=True)
+
+        # The parent id of the query should be same because while moving, a new document is created with old slug
+        # and make redirect to the old document
+        eq_(response.request['QUERY_STRING'], 'slug=children_document&parent={}'.format(doc.id))
```
Comment (jwhitlock) [inline kuma/wiki/tests/test_views.py]: This test breaks if the query string is reversed. The ``parent`` query string parameter should be tested directly.

File: kuma/wiki/tests/test_views.py [hunk 6]
```diff
@@ -1190,6 +1190,22 @@ def test_create_on_404(self):
                                            locale=locale))
         eq_(302, response.status_code)
 
+    def test_creating_child_of_redirect(self):
```
Comment (jwhitlock) [inline kuma/wiki/tests/test_views.py]: A second test is needed for a zoned URL, which often breaks.

File: kuma/wiki/models.py (hunk 2 above)
Comment (jwhitlock) [inline kuma/wiki/models.py]: This function requires unit tests.

File: kuma/wiki/utils.py [hunk 7]
```diff
@@ -53,6 +58,33 @@ def locale_and_slug_from_path(path, request=None, path_locale=None):
     return (locale, slug, needs_redirect)
 
 
+def get_doc_components_from_url(url, required_locale=None, check_host=True):
```
Comment (jwhitlock) [inline kuma/wiki/utils.py]: This function requires tests.  ``required_locale`` and ``check_host`` should be removed or fully tested.

File: kuma/wiki/tests/test_views.py (hunk 4 above)
Comment (safwanrahman) [inline kuma/wiki/tests/test_views.py]: I am planning to change the functionality of moving mage aka redirect page in future. The way its implemented seems to be very strange both kuma and kitsune (Actually, its implemented badly in kitsune, that did not change in kuma era)
So therefore I am thinking to make the redirect with _move_tree
So in future if we change the functionality, the test will not break

**Comment by safwanrahman** [comment]
@jwhitlock Added the tests and updated as you have commented.
r?

**Comment by safwanrahman** [comment]
@jwhitlock Added new test. Now it should not decrease the coverage

File: kuma/wiki/tests/test_views.py [hunk 8]
```diff
@@ -1208,6 +1208,48 @@ def test_create_on_404(self):
...
+        root_doc._move_tree(new_slug="moved_doc")
+
+        zoned_child_full_slug = zoned_doc.url_root + "/" + "children_document"
+        response = self.client.get(zoned_child_full_slug, follow=True)
+        print response.status_code
+
+        # The parent id of the query should be same because while moving, a new document is created with old slug
+        # and make redirect to the old document
+        parameters = parse_qs(response.request['QUERY_STRING'])
+        assert parameters['parent'][0] == str(root_doc.id)
```
Comment (jwhitlock) [inline kuma/wiki/tests/test_views.py]: This works, and tests that ``/en-US/zoned_url/children_doc`` finds the right parent. A second test is needed though for the document that now follows a redirect chain into a zone, something like:

```
root_doc = rev.document
original_slug = root_doc.slug
# zone it, move it around, test new page, etc., then:
# test /en-US/docs/2017-04-21-12203442670/child2
unzoned_child_full_slug = original_slug + "/child2"  
response = self.client.get(unzoned_child_full_slug, follow=True)
# ...check status code, parse parameters, and...
assert parameters['parent'][0] == str(root_doc.id)
```
I manually tested, so this should work.

File: kuma/wiki/tests/test_views.py [hunk 9]
```diff
@@ -1208,6 +1208,48 @@ def test_create_on_404(self):
...
+        # Create a zone of the document
+        zoned_doc = DocumentZone(document=root_doc, url_root="zoned_url")
+        zoned_doc.save()
+
+        # Move the document to new slug
+        root_doc._move_tree(new_slug="moved_doc")
+
+        zoned_child_full_slug = zoned_doc.url_root + "/" + "children_document"
+        response = self.client.get(zoned_child_full_slug, follow=True)
+        print response.status_code
```
Comment (jwhitlock) [inline kuma/wiki/tests/test_views.py]: Remove this print statement, or turn it into an assert

File: kuma/wiki/utils.py (hunk 7 above)
Comment (jwhitlock) [inline kuma/wiki/utils.py]: Now tested in ``test_utils.py``

File: kuma/wiki/tests/test_views.py (hunk 6 above)
Comment (jwhitlock) [inline kuma/wiki/tests/test_views.py]: Now tested, but one more zoned case needed for completeness.

**Comment by safwanrahman** [comment]
@jwhitlock Added one more test as you suggested. r?

**Comment by jwhitlock** [comment]
@safwanrahman thanks for keeping at it though several rounds of reviews!

**Pull request merged**